bdb/bdb_verify.c

```diff
@@ -273,245 +213,582 @@ static void printhex(SBUF2 *sb, int (*lua_callback)(void *, const char *),
+    unsigned char expected_keybuf[18 * 1024];
+    unsigned char verify_keybuf[18 * 1024];
+    int rc = 0;
+    int blobsizes[16];
+    int bloboffs[16];
+    int nblobs = 0;
+    int last = comdb2_time_epochms();
+    blob_buffer_t blob_buf[MAXBLOBS] = {{0}};
 
-            /* check if comdb2sc is killed */
-            if ((now - last) > 1000) {
-                if (dropped_connection(sb)) {
-                    cdata->c_close(cdata);
-                    logmsg(LOGMSG_WARN, "condb2sc connection closed, stopped verify\n");
-                    return rc;
-                }
-            }
+    bdb_state_type *bdb_state = par->bdb_state;
 
-            if (progress_report_seconds &&
-                ((now - last) >= (progress_report_seconds * 1000))) {
-                rc = locprint(sb, lua_callback, lua_params, "!verifying dtastripe %d, did %lld records, %d "
-                                "per second\n",
-                            dtastripe, nrecs,
-                            nrecs_progress / progress_report_seconds);
-                if(rc) return ret; //dropped connection
-                last = now;
-                nrecs_progress = 0;
-                sbuf2flush(sb);
-            }
+    DBT dbt_key = {0};
+    dbt_key.data = keybuf;
+    dbt_key.ulen = sizeof(keybuf);
+    dbt_key.flags = DB_DBT_USERMEM;
+
+    DBT dbt_data = {0};
+    dbt_data.data = databuf;
+    dbt_data.ulen = sizeof(databuf);
+    dbt_data.flags = DB_DBT_USERMEM;
 
-            /* is it the right size? */
-            if (dbt_key.size != sizeof(genid)) {
-                ret = 1;
-                locprint(sb, lua_callback, lua_params, "!bad genid sz %d\n", dbt_key.size);
-                goto next_record;
+    unsigned long long genid;
+    DBT dbt_dta_check_key = {0};
+    dbt_dta_check_key.data = &genid;
+    dbt_dta_check_key.ulen = sizeof(unsigned long long);
+    dbt_dta_check_key.size = sizeof(unsigned long long);
+    dbt_dta_check_key.flags = DB_DBT_USERMEM;
+
+    DBT dbt_dta_check_data = {0};
+    dbt_dta_check_data.data = &verify_keybuf;
+    dbt_dta_check_data.ulen = sizeof(verify_keybuf);
+    dbt_dta_check_data.flags = DB_DBT_USERMEM;
+
+    int atstart = comdb2_time_epochms();
+    int now = atstart;
+    logmsg(LOGMSG_DEBUG, "%p:%s Entering ix=%d\n", (void *)pthread_self(), __func__, ix);
+
+    rc = bdb_state->dbp_ix[ix]->paired_cursor_from_lid(
+        bdb_state->dbp_ix[ix], lid, &ckey, 0);
+    if (rc) {
+        par->verify_status = 1;
+        locprint(par->sb, par->lua_callback, par->lua_params, "!ix %d cursor rc %d\n", ix, rc);
+        return 0;
+    }
+    rc = ckey->c_get(ckey, &dbt_key, &dbt_data, DB_FIRST);
+    if (rc && rc != DB_NOTFOUND) {
+        par->verify_status = 1;
+        locprint(par->sb, par->lua_callback, par->lua_params, "!ix %d first rc %d\n", ix, rc);
+    }
+    while (rc == 0 && !par->client_dropped_connection) {
+        ATOMIC_ADD(par->records_processed, 1);
+        par->nrecs_progress++;
```

Comment:
Double counting records?

Reply:
Yes, just like nrecs++, I guess we want to know how many raw records each of these (data, blobs, indices) is processing so at the end we get ```nrecs = N * nRecords (with N = 1 (data) + B (numblobs) + I (numIdx)) ``` 
Just pointing out that i am not changing that behavior.

Reply:
Sounds good.  As long as the output makes sense.  Want to avoid confusion of someone running select *, seeing 1000 records, then seeing progress reported as 1500 records.

Reply:
I changed the output so that the 'default' way prints exactly what it did before, with same accounting of records (which is really bad if you study it--we zero the nrecs counter and progress counter for every stripe/index and not for blob entries).

Reply:
Also for the parallel mode, we count 'items' instead of records now. A row will be counted multiple times with this method.